# Exploratory Data Analysis

#### Understanding the dataset to explore how the data is present in the database and if there is need of creating aggregated tables that can help with:
- Vendor Selection for profitability
- Product Pricing Optimization

In [2]:
import pandas as pd
import sqlite3

In [3]:
#creating database connection
conn = sqlite3.connect("inventory.db")

In [22]:
# Loop through each table name in the 'tables' DataFrame (column 'name')
for table in tables['name']:
    # Print a visual separator and the name of the table
    print('-'*50, f"{table}", '-'*50)
    
    # Query and print the total number of records in the current table
    print("Count of records:", pd.read_sql(f"SELECT COUNT(*) as count FROM {table}", conn)['count'].values[0])
    
    # Display the first 5 rows of the current table
    display(pd.read_sql(f"SELECT * FROM {table} LIMIT 5", conn))

-------------------------------------------------- purchases --------------------------------------------------
Count of records: 2372474


InventoryId  Store  Brand                   Description   Size  \
0    69_MOUNTMEND_8412     69   8412     Tequila Ocho Plata Fresno  750mL   
1     30_CULCHETH_5255     30   5255  TGI Fridays Ultimte Mudslide  1.75L   
2    34_PITMERDEN_5215     34   5215  TGI Fridays Long Island Iced  1.75L   
3  1_HARDERSFIELD_5255      1   5255  TGI Fridays Ultimte Mudslide  1.75L   
4    76_DONCASTER_2034     76   2034     Glendalough Double Barrel  750mL   

   VendorNumber                   VendorName  PONumber      PODate  \
0           105  ALTAMAR BRANDS LLC               8124  2023-12-21   
1          4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
2          4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
3          4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
4           388  ATLANTIC IMPORTING COMPANY       8169  2023-12-24   

  ReceivingDate InvoiceDate     PayDate  PurchasePrice  Quantity  Dollars  \
0    2024-01-02  2024-01-04  2024-02-16          35.71         6   214.26   
1    2024-01-01  2024-01-07  2024-02-21           9.35         4    37.40   
2    2024-01-02  2024-01-07  2024-02-21           9.41         5    47.05   
3    2024-01-01  2024-01-07  2024-02-21           9.35         6    56.10   
4    2024-01-02  2024-01-09  2024-02-16          21.32         5   106.60   

   Classification  
0               1  
1               1  
2               1  
3               1  
4               1

-------------------------------------------------- purchase_prices --------------------------------------------------
Count of records: 12261


Brand                  Description  Price   Size Volume  Classification  \
0     58  Gekkeikan Black & Gold Sake  12.99  750mL    750               1   
1     62     Herradura Silver Tequila  36.99  750mL    750               1   
2     63   Herradura Reposado Tequila  38.99  750mL    750               1   
3     72         No. 3 London Dry Gin  34.99  750mL    750               1   
4     75    Three Olives Tomato Vodka  14.99  750mL    750               1   

   PurchasePrice  VendorNumber                   VendorName  
0           9.28          8320  SHAW ROSS INT L IMP LTD      
1          28.67          1128  BROWN-FORMAN CORP            
2          30.46          1128  BROWN-FORMAN CORP            
3          26.11          9165  ULTRA BEVERAGE COMPANY LLP   
4          10.94          7245  PROXIMO SPIRITS INC.

-------------------------------------------------- vendor_invoice --------------------------------------------------
Count of records: 5543


VendorNumber                   VendorName InvoiceDate  PONumber  \
0           105  ALTAMAR BRANDS LLC           2024-01-04      8124   
1          4466  AMERICAN VINTAGE BEVERAGE    2024-01-07      8137   
2           388  ATLANTIC IMPORTING COMPANY   2024-01-09      8169   
3           480  BACARDI USA INC              2024-01-12      8106   
4           516  BANFI PRODUCTS CORP          2024-01-07      8170   

       PODate     PayDate  Quantity    Dollars  Freight Approval  
0  2023-12-21  2024-02-16         6     214.26     3.47     None  
1  2023-12-22  2024-02-21        15     140.55     8.57     None  
2  2023-12-24  2024-02-16         5     106.60     4.61     None  
3  2023-12-20  2024-02-05     10100  137483.78  2935.20     None  
4  2023-12-24  2024-02-12      1935   15527.25   429.20     None

-------------------------------------------------- begin_inventory --------------------------------------------------
Count of records: 206529


InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_60      1  HARDERSFIELD     60       Canadian Club 1858 VAP   
2  1_HARDERSFIELD_62      1  HARDERSFIELD     62     Herradura Silver Tequila   
3  1_HARDERSFIELD_63      1  HARDERSFIELD     63   Herradura Reposado Tequila   
4  1_HARDERSFIELD_72      1  HARDERSFIELD     72         No. 3 London Dry Gin   

    Size  onHand  Price   startDate  
0  750mL       8  12.99  2024-01-01  
1  750mL       7  10.99  2024-01-01  
2  750mL       6  36.99  2024-01-01  
3  750mL       3  38.99  2024-01-01  
4  750mL       6  34.99  2024-01-01

-------------------------------------------------- end_inventory --------------------------------------------------
Count of records: 224489


InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_62      1  HARDERSFIELD     62     Herradura Silver Tequila   
2  1_HARDERSFIELD_63      1  HARDERSFIELD     63   Herradura Reposado Tequila   
3  1_HARDERSFIELD_72      1  HARDERSFIELD     72         No. 3 London Dry Gin   
4  1_HARDERSFIELD_75      1  HARDERSFIELD     75    Three Olives Tomato Vodka   

    Size  onHand  Price     endDate  
0  750mL      11  12.99  2024-12-31  
1  750mL       7  36.99  2024-12-31  
2  750mL       7  38.99  2024-12-31  
3  750mL       4  34.99  2024-12-31  
4  750mL       7  14.99  2024-12-31

-------------------------------------------------- sales --------------------------------------------------
Count of records: 12825363


InventoryId  Store  Brand                 Description        Size  \
0  1_HARDERSFIELD_1004      1   1004  Jim Beam w/2 Rocks Glasses       750mL   
1  1_HARDERSFIELD_1004      1   1004  Jim Beam w/2 Rocks Glasses       750mL   
2  1_HARDERSFIELD_1004      1   1004  Jim Beam w/2 Rocks Glasses       750mL   
3  1_HARDERSFIELD_1004      1   1004  Jim Beam w/2 Rocks Glasses       750mL   
4  1_HARDERSFIELD_1005      1   1005     Maker's Mark Combo Pack  375mL 2 Pk   

   SalesQuantity  SalesDollars  SalesPrice   SalesDate  Volume  \
0              1         16.49       16.49  2024-01-01   750.0   
1              2         32.98       16.49  2024-01-02   750.0   
2              1         16.49       16.49  2024-01-03   750.0   
3              1         14.49       14.49  2024-01-08   750.0   
4              2         69.98       34.99  2024-01-09   375.0   

   Classification  ExciseTax  VendorNo                   VendorName  
0               1       0.79     12546  JIM BEAM BRANDS COMPANY      
1               1       1.57     12546  JIM BEAM BRANDS COMPANY      
2               1       0.79     12546  JIM BEAM BRANDS COMPANY      
3               1       0.79     12546  JIM BEAM BRANDS COMPANY      
4               1       0.79     12546  JIM BEAM BRANDS COMPANY

In [17]:
purchases = pd.read_sql_query("select * from purchases where VendorNumber = 4466", conn)
purchases

InventoryId  Store  Brand                   Description   Size  \
0        30_CULCHETH_5255     30   5255  TGI Fridays Ultimte Mudslide  1.75L   
1       34_PITMERDEN_5215     34   5215  TGI Fridays Long Island Iced  1.75L   
2     1_HARDERSFIELD_5255      1   5255  TGI Fridays Ultimte Mudslide  1.75L   
3       38_GOULCREST_5215     38   5215  TGI Fridays Long Island Iced  1.75L   
4     59_CLAETHORPES_5215     59   5215  TGI Fridays Long Island Iced  1.75L   
...                   ...    ...    ...                           ...    ...   
2187     81_PEMBROKE_5215     81   5215  TGI Fridays Long Island Iced  1.75L   
2188   62_KILMARNOCK_5255     62   5255  TGI Fridays Ultimte Mudslide  1.75L   
2189    34_PITMERDEN_5215     34   5215  TGI Fridays Long Island Iced  1.75L   
2190     6_GOULCREST_5215      6   5215  TGI Fridays Long Island Iced  1.75L   
2191    35_HALIVAARA_5255     35   5255  TGI Fridays Ultimte Mudslide  1.75L   

      VendorNumber                   VendorName  PONumber      PODate  \
0             4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
1             4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
2             4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
3             4466  AMERICAN VINTAGE BEVERAGE        8207  2023-12-27   
4             4466  AMERICAN VINTAGE BEVERAGE        8207  2023-12-27   
...            ...                          ...       ...         ...   
2187          4466  AMERICAN VINTAGE BEVERAGE       13595  2024-12-20   
2188          4466  AMERICAN VINTAGE BEVERAGE       13595  2024-12-20   
2189          4466  AMERICAN VINTAGE BEVERAGE       13595  2024-12-20   
2190          4466  AMERICAN VINTAGE BEVERAGE       13595  2024-12-20   
2191          4466  AMERICAN VINTAGE BEVERAGE       13595  2024-12-20   

     ReceivingDate InvoiceDate     PayDate  PurchasePrice  Quantity  Dollars  \
0       2024-01-01  2024-01-07  2024-02-21           9.35         4    37.40   
1       2024-01-02  2024-01-07  2024-02-21           9.41         5    47.05   
2       2024-01-01  2024-01-07  2024-02-21           9.35         6    56.10   
3       2024-01-07  2024-01-19  2024-02-26           9.41         6    56.46   
4       2024-01-05  2024-01-19  2024-02-26           9.41         6    56.46   
...            ...         ...         ...            ...       ...      ...   
2187    2024-12-29  2025-01-04  2025-02-10           9.41         6    56.46   
2188    2024-12-28  2025-01-04  2025-02-10           9.35         5    46.75   
2189    2024-12-28  2025-01-04  2025-02-10           9.41         5    47.05   
2190    2024-12-31  2025-01-04  2025-02-10           9.41         6    56.46   
2191    2024-12-27  2025-01-04  2025-02-10           9.35         5    46.75   

      Classification  
0                  1  
1                  1  
2                  1  
3                  1  
4                  1  
...              ...  
2187               1  
2188               1  
2189               1  
2190               1  
2191               1  

[2192 rows x 16 columns]

In [8]:
vendor_invoice = pd.read_sql_query("select * from vendor_invoice where VendorNumber = 4466", conn)
vendor_invoice.head()

VendorNumber                   VendorName InvoiceDate  PONumber  \
0          4466  AMERICAN VINTAGE BEVERAGE    2024-01-07      8137   
1          4466  AMERICAN VINTAGE BEVERAGE    2024-01-19      8207   
2          4466  AMERICAN VINTAGE BEVERAGE    2024-01-18      8307   
3          4466  AMERICAN VINTAGE BEVERAGE    2024-01-27      8469   
4          4466  AMERICAN VINTAGE BEVERAGE    2024-02-04      8532   

       PODate     PayDate  Quantity  Dollars  Freight Approval  
0  2023-12-22  2024-02-21        15   140.55     8.57     None  
1  2023-12-27  2024-02-26       335  3142.33    16.97     None  
2  2024-01-03  2024-02-18        41   383.35     1.99     None  
3  2024-01-14  2024-03-11        72   673.20     3.30     None  
4  2024-01-19  2024-03-15        79   740.21     3.48     None

In [19]:
sales = pd.read_sql_query("select * from sales where VendorNo = 4466", conn)
sales

InventoryId  Store  Brand                   Description   Size  \
0     1_HARDERSFIELD_5215      1   5215  TGI Fridays Long Island Iced  1.75L   
1     1_HARDERSFIELD_5215      1   5215  TGI Fridays Long Island Iced  1.75L   
2     1_HARDERSFIELD_5215      1   5215  TGI Fridays Long Island Iced  1.75L   
3     1_HARDERSFIELD_5215      1   5215  TGI Fridays Long Island Iced  1.75L   
4     1_HARDERSFIELD_5215      1   5215  TGI Fridays Long Island Iced  1.75L   
...                   ...    ...    ...                           ...    ...   
9448     9_BLACKPOOL_5215      9   5215  TGI Fridays Long Island Iced  1.75L   
9449     9_BLACKPOOL_5255      9   5255  TGI Fridays Ultimte Mudslide  1.75L   
9450     9_BLACKPOOL_5255      9   5255  TGI Fridays Ultimte Mudslide  1.75L   
9451     9_BLACKPOOL_5255      9   5255  TGI Fridays Ultimte Mudslide  1.75L   
9452     9_BLACKPOOL_5255      9   5255  TGI Fridays Ultimte Mudslide  1.75L   

      SalesQuantity  SalesDollars  SalesPrice   SalesDate  Volume  \
0                 1         12.99       12.99  2024-01-09  1750.0   
1                 1         12.99       12.99  2024-01-12  1750.0   
2                 1         12.99       12.99  2024-01-15  1750.0   
3                 1         12.99       12.99  2024-01-21  1750.0   
4                 1         12.99       12.99  2024-01-23  1750.0   
...             ...           ...         ...         ...     ...   
9448              1         12.99       12.99  2024-12-21  1750.0   
9449              1         12.99       12.99  2024-12-02  1750.0   
9450              1         12.99       12.99  2024-12-09  1750.0   
9451              1         12.99       12.99  2024-12-23  1750.0   
9452              1         12.99       12.99  2024-12-31  1750.0   

      Classification  ExciseTax  VendorNo                   VendorName  
0                  1       1.84      4466  AMERICAN VINTAGE BEVERAGE    
1                  1       1.84      4466  AMERICAN VINTAGE BEVERAGE    
2                  1       1.84      4466  AMERICAN VINTAGE BEVERAGE    
3                  1       1.84      4466  AMERICAN VINTAGE BEVERAGE    
4                  1       1.84      4466  AMERICAN VINTAGE BEVERAGE    
...              ...        ...       ...                          ...  
9448               1       1.84      4466  AMERICAN VINTAGE BEVERAGE    
9449               1       1.84      4466  AMERICAN VINTAGE BEVERAGE    
9450               1       1.84      4466  AMERICAN VINTAGE BEVERAGE    
9451               1       1.84      4466  AMERICAN VINTAGE BEVERAGE    
9452               1       1.84      4466  AMERICAN VINTAGE BEVERAGE    

[9453 rows x 14 columns]

In [23]:
# Group the purchases by Brand and PurchasePrice,
# then sum the Quantity and Dollars for each group
purchases.groupby(['Brand', 'PurchasePrice'])[['Quantity', 'Dollars']].sum()

Quantity   Dollars
Brand PurchasePrice                    
3140  11.19              4640  51921.60
5215  9.41               4923  46325.43
5255  9.35               6215  58110.25

In [26]:
sales.groupby(['Brand'])[['SalesQuantity', 'SalesDollars', 'SalesPrice']].sum()

SalesQuantity  SalesDollars  SalesPrice
Brand                                         
3140            3890      50531.10    30071.85
5215            4651      60416.49    41542.02
5255            6096      79187.04    51180.60

- The purchases table contains actual purchase data, including the date of purchase, products (brands) purchased by vendors, the amount paid (in dollars),
and the quantity purchases.
- The purchase price column is derived from the purchas_prices table, which provides product-wise actual and purchase prices. The combination of vendor
and brand is unique in this table.
- The vendor_invoice table aggregates data from the purchases table, summarizing quantity and dollar amounts, along with an additional column for freight.
  This table maintaines uniqueness based on vendor and PO number.
- The sales table captures actual sales transactions, detailing the brands purchased by vendors, the quantity sold, the selling price, and the revenue earned.
-------
As the data we need for analysis is distributed in different tables, we need to create a summary table containing:
- purchases tansactions made by vendor
- sales transaction data
- freight costs for each vendor
- actual product prices from vendors

In [27]:
vendor_invoice.columns

Index(['InventoryId', 'Store', 'Brand', 'Description', 'Size', 'VendorNumber',
       'VendorName', 'PONumber', 'PODate', 'ReceivingDate', 'InvoiceDate',
       'PayDate', 'PurchasePrice', 'Quantity', 'Dollars', 'Classification'],
      dtype='object')

In [46]:
# Query to summarize freight cost by each vendor
freight_summary = pd.read_sql_query("""
    SELECT VendorNumber, SUM(Freight) AS FreightCost
    FROM vendor_invoice
    GROUP BY VendorName
""", conn)
freight_summary

VendorNumber  FreightCost
0              54         0.48
1              60       367.52
2            1703       172.00
3             105        62.39
4             200         6.19
..            ...          ...
124          9622       222.23
125          9625      1933.19
126          3664     30234.42
127          9815     27100.41
128         90058       448.68

[129 rows x 2 columns]

In [49]:
#Query to compute detailed purchase summary per vendor and brand
#Filters out records where PurchasePrice is zero
pd.read_sql_query("""SELECT 
        p.VendorNumber, 
        p.VendorName, 
        p.Brand, 
        p.PurchasePrice, 
        pp.Volume,
        pp.Price AS Actual_Price, 
        SUM(p.Quantity) AS Total_purchases_quantity, 
        SUM(p.Dollars) AS Total_purchase_dollars
    FROM Purchases AS p
    JOIN purchase_prices AS pp
        ON p.Brand = pp.Brand
    WHERE p.PurchasePrice > 0
    GROUP BY p.VendorNumber, p.VendorName, p.Brand
    ORDER BY Total_purchase_dollars
""", conn)

VendorNumber                   VendorName  Brand  PurchasePrice Volume  \
0              7245  PROXIMO SPIRITS INC.          3065           0.71     50   
1              3960  DIAGEO NORTH AMERICA INC      6127           1.47    200   
2              3924  HEAVEN HILL DISTILLERIES      9123           0.74     50   
3              8004  SAZERAC CO INC                5683           0.39     50   
4              9815  WINE GROUP INC                8527           1.32    750   
...             ...                          ...    ...            ...    ...   
10687          3960  DIAGEO NORTH AMERICA INC      3545          21.89   1750   
10688          3960  DIAGEO NORTH AMERICA INC      4261          16.17   1750   
10689         17035  PERNOD RICARD USA             8068          18.24   1750   
10690          4425        MARTIGNETTI COMPANIES   3405          23.19   1750   
10691          1128  BROWN-FORMAN CORP             1233          26.27   1750   

       Actual_Price  Total_purchases_quantity  Total_purchase_dollars  
0              0.99                         1                    0.71  
1              1.99                         1                    1.47  
2              0.99                         2                    1.48  
3              0.49                         6                    2.34  
4              4.99                         2                    2.64  
...             ...                       ...                     ...  
10687         29.99                    138109              3023206.01  
10688         22.99                    201682              3261197.94  
10689         24.99                    187407              3418303.68  
10690         28.99                    164038              3804041.22  
10691         36.99                    145080              3811251.60  

[10692 rows x 8 columns]

In [48]:
sales.columns

Index(['InventoryId', 'Store', 'Brand', 'Description', 'Size', 'SalesQuantity',
       'SalesDollars', 'SalesPrice', 'SalesDate', 'Volume', 'Classification',
       'ExciseTax', 'VendorNo', 'VendorName'],
      dtype='object')

In [57]:
# Query to summarize sales data per vendor and brand
# Aggregates key metrics: total sales dollars, sales quantity, and excise tax
pd.read_sql_query("""SELECT 
        VendorNo, 
        Brand, 
        SUM(SalesDollars) AS Total_sales_dollars,
        SUM(SalesPrice) as TotalSalesPrice,
        SUM(SalesQuantity) AS TotalSalesQuantity, 
        SUM(ExciseTax) AS TotalExciseTax
    FROM Sales
    GROUP BY VendorNo, Brand
""", conn)

VendorNo  Brand  Total_sales_dollars  TotalSalesPrice  \
0             2  90085               665.82           295.92   
1             2  90609               599.76           449.82   
2            60    771               704.53           494.67   
3            60   3979             66871.69         41682.51   
4           105   2529               359.88            59.98   
...         ...    ...                  ...              ...   
11267    173357   2804              6298.60          3194.29   
11268    173357   3666              8996.40          4873.05   
11269    173357   3848               185.94            92.97   
11270    173357   3909             24540.18         14469.21   
11271    201359  90609              1474.41          1324.47   

       TotalSalesQuantity  TotalExciseTax  
0                      18            2.00  
1                      24            0.52  
2                      47           37.01  
3                    3931         7224.06  
4                      12            9.44  
...                   ...             ...  
11267                 140          110.33  
11268                 360          141.19  
11269                   6            4.71  
11270                 982          773.87  
11271                  59            1.24  

[11272 rows x 6 columns]

In [77]:
import time
start = time.time()
vendor_sales_summary = pd.read_sql_query(""" WITH freight_summary AS (
    SELECT 
        VendorNumber, 
        SUM(Freight) AS FreightCost
    FROM vendor_invoice
    GROUP BY VendorNumber
),

purchase_summary AS (
    SELECT 
        p.VendorNumber, 
        p.VendorName, 
        p.Brand,
        p.Description,
        p.PurchasePrice, 
        pp.Volume,
        pp.Price AS ActualPrice,
        SUM(p.Quantity) AS TotalPurchaseQuantity,
        SUM(p.Dollars) AS TotalPurchaseDollars
    FROM Purchases AS p
    JOIN purchase_prices AS pp
        ON p.Brand = pp.Brand
    WHERE p.PurchasePrice > 0
    GROUP BY p.VendorNumber, p.VendorName, p.Brand, p.PurchasePrice, pp.Volume, pp.Price
),

sales_summary AS (
    SELECT 
        VendorNo,
        Brand, 
        SUM(SalesDollars) AS TotalSalesDollars,
        SUM(SalesQuantity) AS TotalSalesQuantity,
        SUM(SalesPrice) as TotalSalesPrice,
        SUM(ExciseTax) AS TotalExciseTax
    FROM Sales
    GROUP BY VendorNo, Brand
)

SELECT 
    ps.VendorNumber,
    ps.VendorName,
    ps.Brand,
    ps.Description,
    ps.PurchasePrice,
    ps.Volume,
    ps.ActualPrice,
    ps.TotalPurchaseQuantity,
    ps.TotalPurchaseDollars,
    ss.TotalSalesQuantity,
    ss.TotalSalesDollars,
    ss.TotalExciseTax,
    fs.FreightCost
FROM purchase_summary ps
LEFT JOIN sales_summary ss 
    ON ps.VendorNumber = ss.VendorNo  AND ps.Brand = ss.Brand
LEFT JOIN freight_summary fs 
    ON ps.VendorNumber = fs.VendorNumber
ORDER BY ps.TotalPurchaseDollars DESC""", conn)

end = time.time()
elapsed_time = (end - start) / 60
print(f"Elapsed time: {elapsed_time} minutes")

Elapsed time: 0.9591975847880045 minutes


In [78]:
vendor_sales_summary

VendorNumber                   VendorName  Brand  \
0              1128  BROWN-FORMAN CORP             1233   
1              4425        MARTIGNETTI COMPANIES   3405   
2             17035  PERNOD RICARD USA             8068   
3              3960  DIAGEO NORTH AMERICA INC      4261   
4              3960  DIAGEO NORTH AMERICA INC      3545   
...             ...                          ...    ...   
10687          9815  WINE GROUP INC                8527   
10688          8004  SAZERAC CO INC                5683   
10689          3924  HEAVEN HILL DISTILLERIES      9123   
10690          3960  DIAGEO NORTH AMERICA INC      6127   
10691          7245  PROXIMO SPIRITS INC.          3065   

                       Description  PurchasePrice Volume  ActualPrice  \
0          Jack Daniels No 7 Black          26.27   1750        36.99   
1            Tito's Handmade Vodka          23.19   1750        28.99   
2                 Absolut 80 Proof          18.24   1750        24.99   
3           Capt Morgan Spiced Rum          16.17   1750        22.99   
4                  Ketel One Vodka          21.89   1750        29.99   
...                            ...            ...    ...          ...   
10687  Concannon Glen Ellen Wh Zin           1.32    750         4.99   
10688  Dr McGillicuddy's Apple Pie           0.39     50         0.49   
10689              Deep Eddy Vodka           0.74     50         0.99   
10690  The Club Strawbry Margarita           1.47    200         1.99   
10691     Three Olives Grape Vodka           0.71     50         0.99   

       TotalPurchaseQuantity  TotalPurchaseDollars  TotalSalesQuantity  \
0                     145080            3811251.60            142049.0   
1                     164038            3804041.22            160247.0   
2                     187407            3418303.68            187140.0   
3                     201682            3261197.94            200412.0   
4                     138109            3023206.01            135838.0   
...                      ...                   ...                 ...   
10687                      2                  2.64                 5.0   
10688                      6                  2.34               134.0   
10689                      2                  1.48                 2.0   
10690                      1                  1.47                72.0   
10691                      1                  0.71                86.0   

       TotalSalesDollars  TotalExciseTax  FreightCost  
0             5101919.51       260999.20     68601.68  
1             4819073.49       294438.66    144929.24  
2             4538120.60       343854.07    123780.22  
3             4475972.88       368242.80    257032.07  
4             4223107.62       249587.83    257032.07  
...                  ...             ...          ...  
10687              15.95            0.55     27100.41  
10688              65.66            7.04     50293.62  
10689               1.98            0.10     14069.87  
10690             143.28           15.12    257032.07  
10691              85.14            4.46     38994.78  

[10692 rows x 13 columns]

This query generates a vendor-wise sales and purchase summary, which is valuable for:
#### Performance Optimization:
- The query involves heavy joins and aggregations on large datasets like sales and purchases.
- Storing the pre-aggregated results avoids repeated expensive computations.
- Helps in analyzing sales, purchases, and pricing for different vendors and brands.
- Future benefits of storing this data for faster dashboarding & reporting.
- Instead of running expensive queries each time, dashboards can fetch data quickly from vendor_sales_summary.

In [79]:
# Checking data types for any inconsistency
vendor_sales_summary.dtypes

VendorNumber               int64
VendorName                object
Brand                      int64
Description               object
PurchasePrice            float64
Volume                    object
ActualPrice              float64
TotalPurchaseQuantity      int64
TotalPurchaseDollars     float64
TotalSalesQuantity       float64
TotalSalesDollars        float64
TotalExciseTax           float64
FreightCost              float64
dtype: object

In [80]:
# Finding any null values
vendor_sales_summary.isnull().sum()

VendorNumber               0
VendorName                 0
Brand                      0
Description                0
PurchasePrice              0
Volume                     0
ActualPrice                0
TotalPurchaseQuantity      0
TotalPurchaseDollars       0
TotalSalesQuantity       178
TotalSalesDollars        178
TotalExciseTax           178
FreightCost                0
dtype: int64

So, We have 178 records with missing values, which means these are unsold products.

In [81]:
vendor_sales_summary['VendorName'].unique()

array(['BROWN-FORMAN CORP          ', 'MARTIGNETTI COMPANIES',
       'PERNOD RICARD USA          ', 'DIAGEO NORTH AMERICA INC   ',
       'BACARDI USA INC            ', 'JIM BEAM BRANDS COMPANY    ',
       'MAJESTIC FINE WINES        ', 'ULTRA BEVERAGE COMPANY LLP ',
       'STOLI GROUP,(USA) LLC      ', 'PROXIMO SPIRITS INC.       ',
       'MOET HENNESSY USA INC      ', 'CAMPARI AMERICA            ',
       'SAZERAC CO INC             ', 'CONSTELLATION BRANDS INC   ',
       'M S WALKER INC             ', 'SAZERAC NORTH AMERICA INC. ',
       'PALM BAY INTERNATIONAL INC ', 'REMY COINTREAU USA INC     ',
       'SIDNEY FRANK IMPORTING CO  ', 'E & J GALLO WINERY         ',
       'WILLIAM GRANT & SONS INC   ', 'HEAVEN HILL DISTILLERIES   ',
       'DISARONNO INTERNATIONAL LLC', 'EDRINGTON AMERICAS         ',
       'CASTLE BRANDS CORP.        ', 'SOUTHERN WINE & SPIRITS NE ',
       'STE MICHELLE WINE ESTATES  ', 'TRINCHERO FAMILY ESTATES   ',
       'MHW LTD                    ', 'W

In [82]:
vendor_sales_summary['Description'].unique()

array(['Jack Daniels No 7 Black', "Tito's Handmade Vodka",
       'Absolut 80 Proof', ..., 'Crown Royal Apple',
       'Concannon Glen Ellen Wh Zin', 'The Club Strawbry Margarita'],
      dtype=object)

In [83]:
#Changing datatype of Volume
vendor_sales_summary['Volume']= vendor_sales_summary['Volume'].astype('float64')

In [84]:
#Removing extra whitespaces
vendor_sales_summary['VendorName']= vendor_sales_summary['VendorName'].str.strip()

In [85]:
#Replacing missing values as 0
vendor_sales_summary.fillna(0, inplace = True)

In [91]:
#Creating new column(feature) named GrossProfit
vendor_sales_summary['GrossProfit']= vendor_sales_summary['TotalSalesDollars'] - vendor_sales_summary['TotalPurchaseDollars']
vendor_sales_summary['GrossProfit']

0        1290667.91
1        1015032.27
2        1119816.92
3        1214774.94
4        1199901.61
            ...    
10687         13.31
10688         63.32
10689          0.50
10690        141.81
10691         84.43
Name: GrossProfit, Length: 10692, dtype: float64

In [92]:
vendor_sales_summary['GrossProfit'].min()

-52002.780000000006

In [97]:
vendor_sales_summary['GrossProfitMargin'] = vendor_sales_summary['GrossProfit'] / vendor_sales_summary['TotalSalesDollars'] * 100
vendor_sales_summary['GrossProfitMargin']

0        25.297693
1        21.062810
2        24.675786
3        27.139908
4        28.412764
           ...    
10687    83.448276
10688    96.436186
10689    25.252525
10690    98.974037
10691    99.166079
Name: GrossProfitMargin, Length: 10692, dtype: float64

In [101]:
vendor_sales_summary['StockTurnover'] = vendor_sales_summary['TotalSalesQuantity'] / vendor_sales_summary['TotalPurchaseQuantity'] * 100
vendor_sales_summary['StockTurnover']

0          97.910808
1          97.688950
2          99.857529
3          99.370296
4          98.355647
            ...     
10687     250.000000
10688    2233.333333
10689     100.000000
10690    7200.000000
10691    8600.000000
Name: StockTurnover, Length: 10692, dtype: float64

In [102]:
vendor_sales_summary['SalestoPurchaseRatio'] = vendor_sales_summary['TotalSalesDollars'] / vendor_sales_summary['TotalPurchaseDollars']
vendor_sales_summary['SalestoPurchaseRatio']

0          1.338647
1          1.266830
2          1.327594
3          1.372493
4          1.396897
            ...    
10687      6.041667
10688     28.059829
10689      1.337838
10690     97.469388
10691    119.915493
Name: SalestoPurchaseRatio, Length: 10692, dtype: float64

In [107]:
cursor = conn.cursor()

In [117]:
# Creating vendor_sales_summary Table

In [112]:
cursor.execute("""CREATE TABLE vendor_sales_summary (
    VendorNumber INT,
    VendorName VARCHAR(100),
    Brand INT,
    Description VARCHAR(100),
    PurchasePrice DECIMAL(10,2),
    ActualPrice DECIMAL(10,2),
    Volume ,
    TotalPurchaseQuantity INT,
    TotalPurchaseDollars DECIMAL(15,2),
    TotalSalesQuantity INT,
    TotalSalesDollars DECIMAL(15,2),
    TotalSalesPrice DECIMAL(15,2),
    TotalExciseTax DECIMAL(15,2),
    FreightCost DECIMAL(15,2),
    GrossProfit DECIMAL(15,2),
    ProfitMargin DECIMAL(15,2),
    StockTurnover DECIMAL(15,2),
    SalesToPurchaseRatio DECIMAL(15,2),
    PRIMARY KEY (VendorNumber, Brand)
)""")

In [113]:
pd.read_sql_query("""select * from vendor_sales_summary""", conn)

Empty DataFrame
Columns: [VendorNumber, VendorName, Brand, Description, PurchasePrice, ActualPrice, Volume, TotalPurchaseQuantity, TotalPurchaseDollars, TotalSalesQuantity, TotalSalesDollars, TotalSalesPrice, TotalExciseTax, FreightCost, GrossProfit, ProfitMargin, StockTurnover, SalesToPurchaseRatio]
Index: []

In [119]:
#Inserting data

In [116]:
vendor_sales_summary.to_sql('vendor_sales_summary', conn, if_exists ='replace', index = False)

10692

In [125]:
pd.read_sql_query("""select * from vendor_sales_summary""", conn)

VendorNumber                VendorName  Brand  \
0              1128         BROWN-FORMAN CORP   1233   
1              4425     MARTIGNETTI COMPANIES   3405   
2             17035         PERNOD RICARD USA   8068   
3              3960  DIAGEO NORTH AMERICA INC   4261   
4              3960  DIAGEO NORTH AMERICA INC   3545   
...             ...                       ...    ...   
10687          9815            WINE GROUP INC   8527   
10688          8004            SAZERAC CO INC   5683   
10689          3924  HEAVEN HILL DISTILLERIES   9123   
10690          3960  DIAGEO NORTH AMERICA INC   6127   
10691          7245      PROXIMO SPIRITS INC.   3065   

                       Description  PurchasePrice  Volume  ActualPrice  \
0          Jack Daniels No 7 Black          26.27  1750.0        36.99   
1            Tito's Handmade Vodka          23.19  1750.0        28.99   
2                 Absolut 80 Proof          18.24  1750.0        24.99   
3           Capt Morgan Spiced Rum          16.17  1750.0        22.99   
4                  Ketel One Vodka          21.89  1750.0        29.99   
...                            ...            ...     ...          ...   
10687  Concannon Glen Ellen Wh Zin           1.32   750.0         4.99   
10688  Dr McGillicuddy's Apple Pie           0.39    50.0         0.49   
10689              Deep Eddy Vodka           0.74    50.0         0.99   
10690  The Club Strawbry Margarita           1.47   200.0         1.99   
10691     Three Olives Grape Vodka           0.71    50.0         0.99   

       TotalPurchaseQuantity  TotalPurchaseDollars  TotalSalesQuantity  \
0                     145080            3811251.60            142049.0   
1                     164038            3804041.22            160247.0   
2                     187407            3418303.68            187140.0   
3                     201682            3261197.94            200412.0   
4                     138109            3023206.01            135838.0   
...                      ...                   ...                 ...   
10687                      2                  2.64                 5.0   
10688                      6                  2.34               134.0   
10689                      2                  1.48                 2.0   
10690                      1                  1.47                72.0   
10691                      1                  0.71                86.0   

       TotalSalesDollars  TotalExciseTax  FreightCost  
0             5101919.51       260999.20     68601.68  
1             4819073.49       294438.66    144929.24  
2             4538120.60       343854.07    123780.22  
3             4475972.88       368242.80    257032.07  
4             4223107.62       249587.83    257032.07  
...                  ...             ...          ...  
10687              15.95            0.55     27100.41  
10688              65.66            7.04     50293.62  
10689               1.98            0.10     14069.87  
10690             143.28           15.12    257032.07  
10691              85.14            4.46     38994.78  

[10692 rows x 13 columns]